File: dsa/data-structures/merge_sort.py
import math

def merge(a, p, q, r):
  nL = q - p + 1
  nR = r - q

  l = [0 for _ in range(nL)]
  r = [0 for _ in range(nR)]

  for i in range (0, nL):
    l[i] = a[p + i]
  for j in range(nR):
    r[j] = a[q + j + 1]

  i = 0
  j = 0
  k = p
  while i < nL and j < nR:
    if l[i] <= r[j]:
      a[k] = l[i]
      i += 1
    else:
      a[k] = r[j]
      j += 1
    k += 1

  while i < nL:
    a[k] = l[i]
    i += 1
    k += 1

  while j < nR:
    a[k] = r[j]
    j += 1
    k += 1

def merge_sort(a, p, r):
  if p >= r:
    return
  
  q = math.floor((p + r) / 2)
  
  merge_sort(a, p, q)
  merge_sort(a, q + 1, r)

  merge(a, p, q, r)

  return a

sorted = merge_sort([3, 2, 1], 0, 2)
print(sorted)

File: clrs/data-structures/merge_sort.js
function merge(values, p, q, r) {
  const left = values.slice(p, q + 1);
  const right = values.slice(q + 1, r + 1);
  let i = 0;
  let j = 0;
  let k = p;

  while (i < left.length && j < right.length) {
    if (left[i] <= right[j]) {
      values[k] = left[i];
      i += 1;
    } else {
      values[k] = right[j];
      j += 1;
    }
    k += 1;
  }

  while (i < left.length) {
    values[k] = left[i];
    i += 1;
    k += 1;
  }

  while (j < right.length) {
    values[k] = right[j];
    j += 1;
    k += 1;
  }
}

function mergeSort(values, p, r) {
  if (p >= r) {
    return values;
  }

  const q = Math.floor((p + r) / 2);
  mergeSort(values, p, q);
  mergeSort(values, q + 1, r);
  merge(values, p, q, r);
  return values;
}

console.log(mergeSort([3, 2, 1], 0, 2));
